pkg/ipam/ippool.go

```diff
@@ -0,0 +1,231 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package ipam
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"sync"
+
+	"github.com/emirpasic/gods/maps/treemap"
+)
+
+type IPPool struct {
+	cidr      string
+	net       *net.IPNet
+	size      int
+	available *treemap.Map    // IntIP is the key, StringIP is value
+	allocated map[string]bool // resource IPAddress is key, value is always true
+	sync.RWMutex
+}
+
+func NewIPPool(cidr string) (*IPPool, error) {
+	_, network, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return nil, err
+	}
+
+	ones, totalbits := network.Mask.Size()
+	size := int(math.Exp2(float64(totalbits-ones))) - 2 // don't count net and broadcast
+	if size < 2 {
+		return nil, errors.New("invalid CIDR Prefix")
+	}
+
+	pool := &IPPool{
+		cidr:      cidr,
+		net:       network,
+		size:      size,
+		available: treemap.NewWithIntComparator(),
+		allocated: make(map[string]bool),
+	}
+	startingIP := ipToInt(pool.net.IP) + 1
+
+	for i := 0; i < pool.size; i++ {
+		intIP := startingIP + i
+		ip := intToIP(startingIP + i).String()
+		pool.available.Put(intIP, ip)
+	}
+
+	// TODO: RecordAvailability(cidr, size)
+
+	return pool, nil
+}
+
+func ipToInt(ip net.IP) int {
+	intIP := ip
+	if len(ip) == 16 {
+		intIP = ip[12:16]
+	}
+
+	return int(binary.BigEndian.Uint32(intIP))
+}
+
+func intToIP(ip int) net.IP {
+	netIP := make(net.IP, 4)
+	binary.BigEndian.PutUint32(netIP, uint32(ip))
+
+	return netIP
+}
+
+func StringIPToInt(stringIP string) int {
```

Comment:
Does this need to be exported?

Reply:
Not in new ipam. In old IPAM it was being used in controller code.

Reply:
The unit tests use it - I left it exported.